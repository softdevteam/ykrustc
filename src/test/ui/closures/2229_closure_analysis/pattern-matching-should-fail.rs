#![feature(capture_disjoint_fields)]
//~^ WARNING: the feature `capture_disjoint_fields` is incomplete
//~| `#[warn(incomplete_features)]` on by default
//~| see issue #53488 <https://github.com/rust-lang/rust/issues/53488>
#![feature(never_type)]

// Should fake read the discriminant and throw an error
fn test1() {
    let x: !;
    let c1 = || match x { };
    //~^ ERROR: use of possibly-uninitialized variable: `x`
}

// Should fake read the discriminant and throw an error
fn test2() {
    let x: !;
    let c2 = || match x { _ => () };
    //~^ ERROR: borrow of possibly-uninitialized variable: `x`
}

// Testing single variant patterns
enum SingleVariant {
    Points(u32)
}

// Should fake read the discriminant and throw an error
fn test3() {
    let variant: !;
    let c = || {
    //~^ ERROR: borrow of possibly-uninitialized variable: `variant`
        match variant {
            SingleVariant::Points(_) => {}
        }
    };
    c();
}

// Should fake read the discriminant and throw an error
fn test4() {
    let variant: !;
    let c = || {
    //~^ ERROR: borrow of possibly-uninitialized variable: `variant`
        match variant {
            SingleVariant::Points(a) => {
                println!("{:?}", a);
            }
        }
    };
    c();
}

fn test5() {
    let t: !;
    let g: !;

    let a = || {
        match g { };
        //~^ ERROR: use of possibly-uninitialized variable: `g`
        let c = ||  {
            match t { };
            //~^ ERROR: use of possibly-uninitialized variable: `t`
        };

        c();
    };

}

// Should fake read the discriminant and throw an error
fn test6() {
    let x: u8;
    let c1 = || match x { };
    //~^ ERROR: use of possibly-uninitialized variable: `x`
    //~| ERROR: non-exhaustive patterns: type `u8` is non-empty
}

fn main() {
    test1();
    test2();
    test3();
    test4();
    test5();
    test6();
}
