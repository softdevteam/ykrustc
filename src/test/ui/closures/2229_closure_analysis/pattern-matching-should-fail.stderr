warning: the feature `capture_disjoint_fields` is incomplete and may not be safe to use and/or cause compiler crashes
  --> $DIR/pattern-matching-should-fail.rs:1:12
   |
LL | #![feature(capture_disjoint_fields)]
   |            ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(incomplete_features)]` on by default
   = note: see issue #53488 <https://github.com/rust-lang/rust/issues/53488> for more information

error[E0004]: non-exhaustive patterns: type `u8` is non-empty
  --> $DIR/pattern-matching-should-fail.rs:72:23
   |
LL |     let c1 = || match x { };
   |                       ^
   |
   = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
   = note: the matched value is of type `u8`

error[E0381]: use of possibly-uninitialized variable: `x`
  --> $DIR/pattern-matching-should-fail.rs:10:23
   |
LL |     let c1 = || match x { };
   |                       ^ use of possibly-uninitialized `x`

error[E0381]: borrow of possibly-uninitialized variable: `x`
  --> $DIR/pattern-matching-should-fail.rs:17:14
   |
LL |     let c2 = || match x { _ => () };
   |              ^^       - borrow occurs due to use in closure
   |              |
   |              use of possibly-uninitialized `x`

error[E0381]: borrow of possibly-uninitialized variable: `variant`
  --> $DIR/pattern-matching-should-fail.rs:29:13
   |
LL |     let c = || {
   |             ^^ use of possibly-uninitialized `variant`
LL |
LL |         match variant {
   |               ------- borrow occurs due to use in closure

error[E0381]: borrow of possibly-uninitialized variable: `variant`
  --> $DIR/pattern-matching-should-fail.rs:41:13
   |
LL |     let c = || {
   |             ^^ use of possibly-uninitialized `variant`
LL |
LL |         match variant {
   |               ------- borrow occurs due to use in closure

error[E0381]: use of possibly-uninitialized variable: `g`
  --> $DIR/pattern-matching-should-fail.rs:57:15
   |
LL |         match g { };
   |               ^ use of possibly-uninitialized `g`

error[E0381]: use of possibly-uninitialized variable: `t`
  --> $DIR/pattern-matching-should-fail.rs:60:19
   |
LL |             match t { };
   |                   ^ use of possibly-uninitialized `t`

error[E0381]: use of possibly-uninitialized variable: `x`
  --> $DIR/pattern-matching-should-fail.rs:72:23
   |
LL |     let c1 = || match x { };
   |                       ^ use of possibly-uninitialized `x`

error: aborting due to 8 previous errors; 1 warning emitted

Some errors have detailed explanations: E0004, E0381.
For more information about an error, try `rustc --explain E0004`.
